commands/migratelxc.go

```diff
@@ -0,0 +1,291 @@
+	}
+	defer st.Close()
+
+	// Collect LXC container machines by host.
+	byHost, err := getLXCContainers(st)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// List LXD containers on each host, so we know which LXC
+	// containers still need migrating. This is required to make
+	// the migrate-lxc command idempotent/resumable.
+	var group errgroup.Group
+	hosts := make([]*state.Machine, 0, len(byHost))
+	lxdContainers := make([]map[string]*api.Container, len(byHost))
+	for host := range byHost {
+		i := len(hosts)
+		hosts = append(hosts, host)
+		host := host // copy for closure
+		group.Go(func() error {
+			containers, err := ListLXDContainers(host)
+			if err != nil {
+				return errors.Annotatef(err,
+					"listing LXD containers for host %q",
+					host.Id(),
+				)
+			}
+			lxdContainers[i] = containers
+			return nil
+		})
+	}
+	if err := group.Wait(); err != nil {
+		return errors.Annotate(err, "listing LXD containers")
+	}
+	lxdByHost := make(map[*state.Machine]map[string]*api.Container)
+	for i, containers := range lxdContainers {
+		lxdByHost[hosts[i]] = containers
+	}
+
+	// Filter out the LXC containers that have already been migrated.
+	environUUID := st.EnvironUUID()
+	namespace, err := instance.NewNamespace(environUUID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	oldNames := make(map[*state.Machine]string) // lxc container -> old name
+	newNames := make(map[string]string)         // old name -> new name
+	for host, containers := range byHost {
+		lxdContainers := lxdByHost[host]
+		nonMigrated := make([]*state.Machine, 0, len(containers))
+		for _, container := range containers {
+			instanceId, err := container.InstanceId()
+			if err != nil {
+				return err
+			}
+			oldName := string(instanceId)
+			oldNames[container] = oldName
+			newName, err := namespace.Hostname(container.Id())
+			if err != nil {
+				return errors.Trace(err)
+			}
+			newNames[oldName] = newName
+			if lxdContainers[oldName] != nil {
+				// migrated but not yet renamed
+				logger.Debugf(
+					"LXC container %q already migrated to LXD (%q)",
+					container.Id(), oldName,
+				)
+				continue
+			}
+			if lxdContainers[newName] != nil {
+				// migrated and renamed
+				logger.Debugf(
+					"LXC container %q already migrated to LXD (%q)",
+					container.Id(), newName,
+				)
+				continue
+			}
+			nonMigrated = append(nonMigrated, container)
+		}
+		if len(nonMigrated) == 0 {
+			delete(byHost, host)
+		} else {
+			byHost[host] = nonMigrated
+		}
+	}
+
+	// Make sure the non-migrated LXC containers are stopped.
+	for host, containers := range byHost {
+		for _, container := range containers {
+			logger.Debugf("stopping LXC container %q", container.Id())
+			host, container := host, container // copy for closure
+			group.Go(func() error {
+				return errors.Annotatef(
+					StopLXCContainer(container, host),
+					"stopping LXC container %q",
+					container.Id(),
+				)
+			})
+		}
+	}
+	if err := group.Wait(); err != nil {
```

Comment:
I was surprised by reusing the group like this, but I guess if there had been any errors in previous phases we would have returned already.